File: plugins/quotes.1h.py
#!/usr/bin/env python3
"""Daily Quote - Shows an inspirational quote"""
import json
import urllib.request
import random

QUOTES = [
    {"text": "The only way to do great work is to love what you do.", "author": "Steve Jobs"},
    {"text": "Innovation distinguishes between a leader and a follower.", "author": "Steve Jobs"},
    {"text": "Stay hungry, stay foolish.", "author": "Steve Jobs"},
    {"text": "Code is like humor. When you have to explain it, it's bad.", "author": "Cory House"},
    {"text": "First, solve the problem. Then, write the code.", "author": "John Johnson"},
    {"text": "Simplicity is the soul of efficiency.", "author": "Austin Freeman"},
    {"text": "Make it work, make it right, make it fast.", "author": "Kent Beck"},
    {"text": "Talk is cheap. Show me the code.", "author": "Linus Torvalds"},
    {"text": "Any fool can write code that a computer can understand.", "author": "Martin Fowler"},
    {"text": "The best error message is the one that never shows up.", "author": "Thomas Fuchs"},
]

def get_quote():
    try:
        url = "https://api.quotable.io/random"
        with urllib.request.urlopen(url, timeout=3) as response:
            data = json.loads(response.read().decode())
            return {"text": data['content'], "author": data['author']}
    except:
        return random.choice(QUOTES)

quote = get_quote()
text = quote['text']
author = quote['author']

# Truncate for display
display_text = text[:40] + "..." if len(text) > 40 else text

print(f" {display_text}")
print("---")
print(f'"{text}"')
print(f"  - {author}")
print("---")
print("New Quote | refresh=true")

File: plugins/quotes/quotes.1h.lua
-- quotes.1h.lua
-- Daily Quote via Crossbar web

local function env(name, default)
    local value = crossbar.env(name, default)
    if value == nil or value == '' then
        return default
    end
    return value
end

local function env_num(name, default)
    local value = env(name, tostring(default))
    local num = tonumber(value)
    if num == nil then
        return default
    end
    return num
end

local max_len = env_num('QUOTE_MAX_LEN', 40)
if max_len < 20 then max_len = 20 end
if max_len > 300 then max_len = 300 end
local response, err = crossbar.web('https://api.quotable.io/random')
if response == nil then
    print('QUOTE Error | color=gray')
    print('---')
    print(err or 'Failed to fetch quote')
    print('---')
    print('Refresh | refresh=true')
    return
end

if response.error or (response.status and response.status >= 400) then
    print('QUOTE Error | color=gray')
    print('---')
    print(response.message or ('HTTP ' .. tostring(response.status)))
    print('---')
    print('Refresh | refresh=true')
    return
end

local data = response.data
local text = nil
local author = nil
if type(data) == 'table' then
    text = data.content
    author = data.author
end

if text == nil then
    print('QUOTE Error | color=gray')
    print('---')
    print('Failed to fetch quote')
    print('---')
    print('Refresh | refresh=true')
    return
end

local display_text = text
if #display_text > max_len then
    display_text = string.sub(display_text, 1, max_len) .. '...'
end

print('QUOTE ' .. display_text)
print('---')
print('"' .. text .. '"')
if author and author ~= '' then
    print('  - ' .. author)
end
print('---')
print('Refresh | refresh=true')
